### ISSUE TITLE
BullMQ Pro Job `timeout` usage

### ISSUE THREAD

**Issue by lukepolo** [taskforcesh/bullmq#1393]
Im able to use the `ttl` on the worker, but not all the jobs on the worker need the `ttl`. I see the `timeout` on the job but it doesn't seem to work . 

Is there a way to TTL per job without doing this

```
const worker = new WorkerPro(queueName, processor, {
  ttl: { test1: 100, test2: 200 },
  connection,
});
```

our workers / jobs are decoupled for the most part .

**Comment by lukepolo** [comment]
also an update, if you specify a ttl on a single job name, the rest get some sort of default 


![image](https://user-images.githubusercontent.com/2066668/186907322-025d4571-802d-47f3-a963-68ceeb2be24d.png)

This causes all other jobs to fail

**Comment by manast** [comment]
> Is there a way to TTL per job without doing this
> 
> ```
> const worker = new WorkerPro(queueName, processor, {
>   ttl: { test1: 100, test2: 200 },
>   connection,
> });
> ```
> 
> our workers / jobs are decoupled for the most part .

You can just use a TTL for all your jobs: https://docs.bullmq.io/bullmq-pro/observables/cancelation

**Comment by JustDoItSascha** [comment]
@manast @lukepolo I have the same question like the original author and it seems like the question is not properly answered:

What is the timeout property in the job options for when not to let the job fail with a timeout error?

**Comment by manast** [comment]
@JustDoItSascha Can you rephrase your question?, the double negative is confusing me :)

**Comment by JustDoItSascha** [comment]
@manast Sure :)

When I add a job to a queue, I have the possibility to give this job a `timeout` option. What exactly does this timeout option? Because it's not doing what the OP and I thought it would do.

I would have assumed that if I give it a timeout of `5000` then when the jobs is longer than 5 seconds in active state, it will fail...

**Comment by manast** [comment]
Are you using the Pro version of BullMQ ?

**Comment by JustDoItSascha** [comment]
Yes

**Comment by manast** [comment]
Then it should work, can you please write a snippet where it doesn't?

**Comment by JustDoItSascha** [comment]
I hope this helps. I'm using Bull inside NestJS, but without the default bull package, because it's not the pro version.

```
onModuleInit(): any {
        const connection = {
            host: this.config.get('REDIS_HOST'),
            port: this.config.get('REDIS_PORT'),
            password: this.config.get('REDIS_PASSWORD'),
        };

        const queueScheduler = new QueueSchedulerPro('timeoutTest', {
            autorun: true,
            stalledInterval: 30000,
            connection,
        });

        const worker = new WorkerPro<any, any, string>(
            'timeoutTest',
            async job => await this.processor(job),
            {
                connection,
                concurrency: 1,
                prefix: `bullmqtest`,
            },
        );

        const queue = new QueuePro<any, any, string>(
            'timeoutTest',
            {
                connection,
                prefix: `bullmqtest`,
                defaultJobOptions: {
                    attempts: 2,
                    backoff: {
                        type: 'fixed',
                        delay: 10000,
                    },
                },
            },
        );

        queue.add('test', [], {
            timeout: 1000,
        });
    }

    async processor(job: JobPro<any>) {
        await firstValueFrom(timer(5000));
    }
```

**Comment by manast** [comment]
@JustDoItSascha It seems that you have used the wrong option, it should be on the Worker:
```ts
const worker = new WorkerPro(queueName, processor, {
  ttl: 100,
  connection,
});
```

**Comment by JustDoItSascha** [comment]
@manast But what is the timeout option in the job for? I mean it exists so it must have any purpose?

**Comment by manast** [comment]
We only support TTL on the worker, and requires to implement the job as an observable. The timeout option is deprecated in BullMQ.

**Comment by JustDoItSascha** [comment]
@manast I hope that you don't take this personally, your library does amazing things. But your documentation is close to the worst I've ever seen :D

If I look into the typings, I don't see any deprecated flag:
<img width="630" alt="image" src="https://user-images.githubusercontent.com/8389022/189656725-09399b55-0563-4568-8849-9fbd926a5850.png">

And deprecated does not mean "not working". If something is deprecated then it should still work and will be removed in the next major release. The documentation is a total mess on things like that. :(

**Comment by manast** [comment]
@JustDoItSascha which version of BullMQ are you using? I cannot find that option on our API documentation: https://api.docs.bullmq.io/
Nor on current source code...

**Comment by manast** [comment]
Ok, I see that in the Pro versión there is a timeout option on the job options interface. It must have gotten here by mistake. Do you know where it comes from @roggervalf ? https://github.com/taskforcesh/bullmq-pro/blame/74a4a0ba01f3a447f9dc24f5bbb898bc6afaeaa6/src/interfaces/jobs-pro-options.ts#L12

**Comment by manast** [comment]
@JustDoItSascha Regarding the documentation. Do you have any other examples of things that could be improved apart from the timeout mistake?

**Comment by JustDoItSascha** [comment]
@manast Yes, for example the whole QueueScheduler thing. It is a very complex architecture which doesn't feel "natural" and is explained with like 2 paragraphs.

It would be good, if things are explained in detail with examples what will happen under the hood.

For example: "When a job is in an active state, i.e., it is being processed by a worker, it needs to continuously update the queue to notify that the worker is still working on the job."

What does that mean? Who has to notify the queue? The worker itself? This job processor? How often, what does "continuosly" means? How does the QueueScheduler recognize that a job is stalled?

To have answers to these questions would help to understand the whole library and to understand for example the purpose of Schedulers.

Or why do I have to use an observable when I want to use ttl? As a developer I see no reason why this shouldn't work also with Promises, so it would be nice to have a explanation for this. Or why do I have to return a named function in this Observable. Would an anonymous arrow function also be ok? Why don't you wrap the processor in an observable by yourself in the first place? Why do I have to make it? :D

And then the whole timeout part is under a section named "Observable" and then "Cancellation". Why isn't it under "Job" -> "Timeout" where it belongs? Instead of structuring the documentation in the way a developer will use your library, you structured your documentation in the way you have solved common problems: 

* I don't want to know, what a QueueScheduler is, I want to know how to resume jobs ofter my microservice has restarted. Never ever I would look into QueueScheduler.
* There are two Flows sections.
* Real Time Updates is just empty.
* Sender - Receiver also.
* The Api Doc from TypeDoc is very chaotic. This was absolutely not the best choice...

I hope this helps! Thanks for your open ear!

**Comment by JustDoItSascha** [comment]
Oh and look here:
<img width="817" alt="image" src="https://user-images.githubusercontent.com/8389022/189716867-ed81b3a7-c280-4b91-b1b8-5e81cec1f555.png">

No explanation what group is. And click on [interfaces/jobs-pro-options.ts:4](https://github.com/taskforcesh/bullmq-pro/blob/74a4a0b/src/interfaces/jobs-pro-options.ts#L4) , you will get a 404...
